### PR TITLE
Add Hardware Key login audit event fields

### PR DESCRIPTION
**Pull request by Joerger**
Adds audit event fields to track hardware key logins.

Audit log snippets:
```json
{
  "code": "T1000I",
  "ei": 0,
  "event": "user.login",
  "required_private_key_policy": "hardware_key_touch",
}
```

```json
{
  "code": "TC000I",
  "ei": 0,
  "event": "cert.create",
  "identity": {
    "private_key_policy": "hardware_key_touch",
  },
}
```

```json
{
  "code": "T2006I",
  "event": "session.data",
  "private_key_policy": "hardware_key_touch",
}
```

### CODE REVIEW

File: lib/auth/methods.go [hunk 1]
```diff
@@ -136,6 +127,30 @@ func (s *Server) AuthenticateUser(ctx context.Context, req AuthenticateUserReque
...
+
+	switch {
+	case username != "" && actualUsername != "" && username != actualUsername:
+		log.Warnf("Authenticate user mismatch (%q vs %q). Using request user (%q)", username, actualUsername, username)
+	case username == "" && actualUsername != "":
+		log.Debugf("User %q authenticated via passwordless", actualUsername)
+		username = actualUsername
+	}
+
+	event.Code = events.UserLocalLoginCode
```
Comment (r0mant) [inline lib/auth/methods.go:152]: Looks like half of this function's logic now is to build fields for the audit events. Can it be factored out into a separate method so the function looks like this?
```go
func (s *Server) AuthenticateUser() xxx {
    xxx, err := s.authenticateUser(xxx)
    // handle err

    xxx := s.emitAuthAuditEvent(yyy)
    // ...
}
```

File: lib/srv/desktop/audit.go [hunk 2]
```diff
@@ -47,11 +47,28 @@ type desktopSessionAuditor struct {
 	auditCache sharedDirectoryAuditCache
 }
 
+func (d *desktopSessionAuditor) makeSessionMetadata() events.SessionMetadata {
```
Comment (ibeckermayer) [inline lib/srv/desktop/audit.go:50]: Possible nit: why is this `makeSessionMetadata` while the similar function for `*ServerContext` is `GetSessionMetadata`?

File: lib/auth/methods.go [hunk 3]
```diff
@@ -126,61 +137,99 @@ func (s *Server) AuthenticateUser(ctx context.Context, req AuthenticateUserReque
...
+
+	clusterName, err := s.GetClusterName()
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	accessInfo := services.AccessInfoFromUserState(userState)
+	checker, err := services.NewAccessChecker(accessInfo, clusterName.GetClusterName(), s)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
```
Comment (ibeckermayer) [inline lib/auth/methods.go:165]: Was it a bug that we weren't doing this check in `AuthenticateWebUser` previously?
https://github.com/gravitational/teleport/blob/70de43e913c0ec10c8d3e55d1cbceefa68a6d87b/lib/auth/methods.go#L490

Separate but related: generally it strikes me that this function is not very well tested (afaict) despite being load bearing for system security.

**Review by ibeckermayer — APPROVED** [review]

File: api/proto/teleport/legacy/types/events/events.proto [hunk 4]
```diff
@@ -59,6 +59,8 @@ message SessionMetadata {
   string SessionID = 1 [(gogoproto.jsontag) = "sid"];
   // WithMFA is a UUID of an MFA device used to start this session.
   string WithMFA = 2 [(gogoproto.jsontag) = "with_mfa,omitempty"];
+  // PrivateKeyPolicy is the private key policy of the private key used to start this session.
+  string private_key_policy = 3;
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/proto/teleport/legacy/types/events/events.proto:63]: Not recent, this is the ideal way we should be doing it and it moves us away from `gogoproto`. If you take a look at our newer proto files, like `joinservice.proto`, you'll see that all fields look like this. IMO it's fine to mix the new format in with the old proto files.

File: lib/auth/methods.go (hunk 3 above)
Comment (Joerger) [reply lib/auth/methods.go:165]: That we weren't getting the `checker` inside the function? It wasn't a bug, we were just getting it after the function call. This change is just a refactor.

I think most paths are covered by `auth_login_test.go`.

**Review by r0mant — COMMENTED** [review]
Overall lgtm bar the proto field issue (unless it's the preferred way of doing things now).

File: api/proto/teleport/legacy/types/events/events.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply api/proto/teleport/legacy/types/events/events.proto:63]: I follow the code around - if the message does UpperCamel, I do that. If the message is brand new, then I follow the regular proto style (lower_snake). Mixing is weird for sure.

**Review by r0mant — APPROVED** [review]

File: lib/auth/methods.go (hunk 3 above)
Comment (ibeckermayer) [reply lib/auth/methods.go:165]: Yeah my question makes no sense, we aren't even using the checker to check anything here. My mistake.

**Pull request merged**